### PR TITLE
[ENG-2378] feat: add netsuite restapi module

### PR DESCRIPTION
**Pull request by RajatPawar**
#### Listing records (internally) which are then fetched 
Screenshot 2025-07-28 at 2 22 57 PM
Screenshot 2025-07-28 at 2 23 05 PM

### Read-write 
Screenshot 2025-07-28 at 3 22 06 PM

### Metadata 
Screenshot 2025-07-28 at 2 04 45 PM
Screenshot 2025-07-28 at 2 04 27 PM
Screenshot 2025-07-28 at 2 04 23 PM

### Delete
Screenshot 2025-07-28 at 3 44 22 PM

### Errors
Screenshot 2025-07-28 at 3 32 00 PM

### CODE REVIEW

File: providers/netsuite/internal/restapi/read.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+	// Collect the results.
+	results := make([]map[string]any, len(recordsToFetch))
+
+	var allErrors []error
+
+	for res := range resultChan {
+		if res.err != nil {
+			allErrors = append(allErrors, res.err)
+		} else {
+			results[res.index] = res.data
```
Comment (Cobalt0s) [inline providers/netsuite/internal/restapi/read.go:191]: Given the line:
```
results := make([]map[string]any, **len(recordsToFetch))**
```

When the error occurs the value won't be placed into the array "slot". Therefore, they will be set to `nil`. I checked downstream and no nil pointer can occur. Also the overall procedure would collect all errors and disregard rows data.

Predeclaring the slice with a size implies that each slot will be initialized. Otherwise, it is better to use append and for performence -- set capacity instead.

I think if `ReadResult` had both error and rows then the comment I have above would be more critical.

**Review by Cobalt0s — APPROVED** [review]
Looks good!
I’ve left some notes.

Also, it seems there are quite a few edge cases in how NetSuite works. Let me know if I can help add some mock tests in a follow-up PR. Also, ping me on Slack with your thoughts about the adapter “pattern.”

File: providers/netsuite.go [hunk 2]
```diff
@@ -1,8 +1,16 @@
...
-const Netsuite Provider = "netsuite"
+const (
+	Netsuite Provider = "netsuite"
 
-// nolint:lll
+	// NetsuiteModuleSuiteQL is a read-only module that uses SuiteQL to read data.
+	NetsuiteModuleSuiteQL = "suiteql"
+
+	// NetsuiteModuleRESTAPI is a read-write module that uses the REST API to read and write data.
+	NetsuiteModuleRESTAPI = "restapi"
```
Comment (laurenzlong) [inline providers/netsuite.go:10]: what do you think of `rest` as the module name instead? it's simpler and reads better than `restapi`, which looks like it should be pronounced "res-tah-pee"

Comment (laurenzlong) [reply providers/netsuite.go:10]: similarly the constant name can be `NetsuiteModuleREST` for brevity

File: providers/netsuite.go [hunk 3]
```diff
@@ -28,6 +36,24 @@ func init() {
 			Subscribe: false,
 			Write:     false,
 		},
+		DefaultModule: NetsuiteModuleRESTAPI,
+		Modules: &Modules{
+			NetsuiteModuleSuiteQL: {
+				DisplayName: "Netsuite (SuiteQL)",
+				BaseURL:     "https://{{.workspace}}.suitetalk.api.netsuite.com/services/rest/query",
+				Support: Support{
```
Comment (laurenzlong) [inline providers/netsuite.go:44]: isn't proxy supported for one or both of these modules?

File: providers/netsuite/connector.go [hunk 4]
```diff
@@ -20,30 +20,85 @@ type Connector struct {
 	common.RequireAuthenticatedClient
 	common.RequireWorkspace
 
-	// Supported operations
-	components.SchemaProvider
+	// TODO: Expose concurrency knobs to the server.
```
Comment (laurenzlong) [inline providers/netsuite/connector.go:23]: a little confused by this comment, can you elaborate a bit more?

File: providers/netsuite/internal/restapi/metadata.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+		}
+
+		// backward compatibility
+		result.FieldsMap[name] = displayName
+	}
+
+	return result, nil
+}
+
+func oneOf(candidates ...string) string {
```
Comment (laurenzlong) [inline providers/netsuite/internal/restapi/metadata.go:93]: `oneOf` is a confusing name for this function, maybe `firstNonEmpty` or something

File: providers/netsuite/internal/suiteql/read.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	"github.com/amp-labs/connectors/providers/netsuite/internal/shared"
+)
+
+const (
+	// SuiteQL timestamp format for TO_TIMESTAMP function. This seems to be a magic
+	// date format that is accepted by SuiteQL. We should verify it this actually
+	// works for all instances.
+	suiteQLTimestampFormat = "2006-01-02 15:04:05.000000000"
+
+	defaultLimit = 1000
```
Comment (laurenzlong) [inline providers/netsuite/internal/suiteql/read.go:23]: `maxRecords` or `pageSize` would be a clearer name, I had to look at how it's used in the code to understand what it actually means

**Review by laurenzlong — APPROVED** [review]

File: providers/netsuite/internal/suiteql/metadata.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+	for field := range record {
+		fields[field] = common.FieldMetadata{DisplayName: field}
+		fieldsMap[field] = field
+	}
+
+	metadata := &common.ObjectMetadata{
+		DisplayName: objectName,
+		Fields:      fields,
+		FieldsMap:   fieldsMap,
```
[the remark this answers is not in the document]
Comment (RajatPawar) [reply providers/netsuite/internal/suiteql/metadata.go:71]: cool, fixed! Updated to 
```

	fields := common.FieldsMetadata{}

	for field := range response.Items[0] {
		fields.AddFieldWithDisplayOnly(field, field)
	}

	return common.NewObjectMetadata(objectName, fields), nil
```
